### PR TITLE
unban+admin

### PR DESCRIPTION
**Pull request by substack**
This patch builds on #78 and #77 and adds:

* unban
* addMod
* removeMod
* addAdmin
* removeAdmin
* moderation.getRole
* moderation.modInfo
* moderation.listMods
* moderation.listAdmins

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,11 +109,80 @@ This means that if a group of people all specify the same *moderation key*,
...
+- `reason` (string, default="") - reason for the banishment.
+
+### cabal.unban(key, [opts], [cb])
+
+Unban a user by their public key.
+
+Valid `opts` include:
+
+- `channel` (string, default="@") - the channel to ban the user from. `"@"` is cabal-wide.
+- `reason` (string, default="") - reason for the banishment.
```
Comment (cblgh) [inline README.md:128]: nitpick: should be `unban`

File: README.md [hunk 2]
```diff
@@ -109,11 +109,80 @@ This means that if a group of people all specify the same *moderation key*,
...
+
+### cabal.addMod(key, [opts], [cb])
+
+Add a moderator by their `key`.
+
+Moderators can ban and unban users who are not admins or moderators.
+
+Valid `opts` include:
+
+- `channel` (string, default="@") - the channel to ban the user from. `"@"` is cabal-wide.
```
Comment (cblgh) [inline README.md:138]: does the underlying view allow adding a mod/admin for a single channel?

nitpick: says ban. same case for the next four segments containing `channel`

File: index.js [hunk 3]
```diff
@@ -179,6 +179,121 @@ Cabal.prototype.publishChannelTopic = function (channel, topic, cb) {
...
+ * @param {string} [opts.channel="@"] - the channel to ban the user from.
+ *                                      (default=cabal-wide)
+ * @param {string} [opts.reason=""] - reason for the banishment.
+ * @param {function} [cb] - callback upon completion.
+ */
+Cabal.prototype.removeAdmin = function (key, opts, cb) {
+  this._modCmd('admin/remove', key, opts, cb)
+}
+
+Cabal.prototype._modCmd = function (type, key, opts, cb) {
```
Comment (cblgh) [inline index.js:260]: i like this structure. i was considering something similar for parsing the extended moderation syntax in `cabal-client`

**Comment by cblgh** [comment]
looks like the failing ci requires pushing `package-lock.json`, to make sure the `materialized-group-auth` version is `^1.2.0`

**Comment by substack** [comment]
I am very hesistant to make any changes to a package-lock in a PR. This creates automatic merge conflicts between any other open PRs.

**Comment by cblgh** [comment]
cool, that makes sense. i'll try to remember to make that part of my workflow going forward, too

File: index.js (hunk 3 above)
Comment (noffle) [reply index.js:260]: Nice!

File: index.js [hunk 4]
```diff
@@ -216,14 +331,17 @@ Cabal.prototype._removeConnection = function (key) {
...
-
 function generateKeyHex () {
   return crypto.keyPair().publicKey.toString('hex')
 }
 
 function isHypercoreKey (key) {
-  if (typeof key === 'string') return key.length === 64 && /^[0-9a-f]+$/.test(key)
-  else if (Buffer.isBuffer(key)) return key.length === 32
+  if (typeof key === 'string') return /^[0-9a-f]{64}$/.test(key)
+  else if (Buffer.isBuffer(key)) return key.toString('hex').length === 32
```
Comment (noffle) [inline index.js:340]: Glad you noticed this bug.

File: views/moderation.js [hunk 5]
```diff
@@ -134,11 +187,16 @@ module.exports = function (cabal, modKey, db) {
           })
           if (--pending === 0) cb(null, banned)
         })
+      },
+
+      getRole: function (core, opts, cb) {
```
Comment (noffle) [inline views/moderation.js:192]: Great API expansion! Do you have capacity to add these to the README docs as you go?

**Comment by substack** [comment]
closing in favor of #83

**Pull request closed without merging**